state/prune.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/juju/errors"
+	"gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// pruneCollection removes collection entries until
+// only entries newer than <maxLogTime> remain and also ensures
+// that the collection is smaller than <maxLogsMB> after the
+// deletion.
+func pruneCollection(st *State, maxHistoryTime time.Duration, maxHistoryMB int, collectionName string, ageField string, timeUnit TimeUnit) error {
+
+	// NOTE(axw) we require a raw collection to obtain the size of the
+	// collection. Take care to include model-uuid in queries where
+	// appropriate.
+	entries, closer := st.getRawCollection(collectionName)
+	defer closer()
+
+	p := collectionPruner{
+		st:       st,
+		coll:     entries,
+		maxAge:   maxHistoryTime,
+		maxSize:  maxHistoryMB,
+		ageField: ageField,
+		timeUnit: timeUnit,
+	}
+	if err := p.validate(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := p.pruneByAge(); err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(p.pruneBySize())
+}
+
+const historyPruneBatchSize = 1000
+const historyPruneProgressSeconds = 15
+
+type doneCheck func() (bool, error)
+
+type TimeUnit string
+
+const (
+	NanoSeconds TimeUnit = "nanoseconds"
+	GoTime      TimeUnit = "goTime"
+)
+
+type collectionPruner struct {
+	st   *State
+	coll *mgo.Collection
+
+	maxAge  time.Duration
+	maxSize int
+
+	ageField string
+	timeUnit TimeUnit
+}
+
+func (p *collectionPruner) validate() error {
+	if p.maxSize < 0 {
+		return errors.NotValidf("non-positive max size")
+	}
+	if p.maxAge < 0 {
+		return errors.NotValidf("non-positive max age")
+	}
+	if p.maxSize == 0 && p.maxAge == 0 {
+		return errors.NotValidf("backlog size and age constraints are both 0")
+	}
+	return nil
+}
+
+func (p *collectionPruner) pruneByAge() error {
+	if p.maxAge == 0 {
+		return nil
+	}
+
+	t := p.st.clock.Now().Add(-p.maxAge)
+	var age interface{}
+
+	if p.timeUnit == NanoSeconds {
```

Comment:
It is needed since some collections marshal to Unix nanoseconds on their way to the db while other use the default Go's `time.Time` marshaling. Both which yield  different results when unmarshaled.